src/crypto/keccak.c

```diff
@@ -99,7 +99,7 @@ void keccak(const uint8_t *in, size_t inlen, uint8_t *md, int mdlen)
     }
     
     // last block and padding
-    if (inlen >= sizeof(temp) || inlen > rsiz || rsiz - inlen + inlen + 1 >= sizeof(temp) || rsiz == 0 || rsiz - 1 >= sizeof(temp) || rsizw * 8 > sizeof(temp))
+    if (inlen + 1 >= sizeof(temp) || inlen > rsiz || rsiz - inlen + inlen + 1 >= sizeof(temp) || rsiz == 0 || rsiz - 1 >= sizeof(temp) || rsizw * 8 > sizeof(temp))
```

Comment:
Doesn't `rsiz - inlen + inlen + 1` simplify to `rsiz + 1`?

Reply:
Yes. It is that way to match "memset(temp + inlen, 0, rsiz - inlen);" below so it's clear what the check is meant for. I expect compilers will generate the same code.

Reply:
Can we drop the `inlen > rsiz` check? It seems impossible after the loop above has finished, which runs until `inlen >= rsiz` is false.

Also, should the comment `// last block and padding` move down, to the next line after the check?

Reply:
All of these should be impossible, that's the point of the check: ensuring that what we thought is true is before we dereference something which assumes it is.

Reply:
I'm a bit worried about [line 90](https://github.com/monero-project/monero/blob/master/src/crypto/keccak.c#L90). `200 - 2 * mdlen` can theoretically be zero or negative. Should we catch this in the check above?

Reply:
Maybe `mdlen > 200` should become `mdlen > 100` in the check above?

Reply:
Thanks, fixed.